[WIP] Infrastructure for extracting syntax tests.

ekpyron opened the pull request:
Refs #3644.

Copies some example test contracts from SolidityNameAndTypeResolution.cpp in separate files in test/libsolidity/syntaxTests/ in which the test expectations are encoded in comments, adds a parser for this encoding and populates the boost test tree with these tests.

ekpyron committed:
Infrastructure for extracting syntax tests in separate test files.

chriseth commented on test/libsolidity/syntaxTests/double_stateVariable_declaration.sol:
```diff
@@ -0,0 +1,8 @@
+// double_stateVariable_declaration.sol
```
Do we need to replicate the file name inside the file?

chriseth commented on test/TestHelper.cpp:
```diff
@@ -43,6 +43,11 @@ Options::Options()
 			ipcPath = suite.argv[i + 1];
 			i++;
 		}
+		else if (string(suite.argv[i]) == "--testpath" && i + 1 < suite.argc)
```
Please also modify the CI scripts accordingly.

ekpyron replied:
tests.sh, isolate_tests.py and appveyor.xml are adjusted

chriseth replied:
I'm not sure if `testpath` is the best way to name this option, but I also cannot think of a better one :)

ekpyron replied:
``testroot``maybe?

chriseth replied:
It's not really about the tests, it is rather about the test data files.

ekpyron replied:
True.

``--testdata``, ``--testdataroot`` or ``--testdatapath`` then :-)? (or hyphenated as ``--test-data-root``, etc.)

At the moment the argument expected for --testpath is the root test directory - libsolidity/syntaxTests is appended to that. One could create a different directory tree under test/, e.g. test/data/, and put all test data files in there - then ``--test-data-path`` or an alternative would probably make more sense.

chriseth replied:
Hm, perhaps it is fine as it is :)

chriseth commented on test/libsolidity/syntaxTests/double_variable_declaration_050.sol:
```diff
@@ -0,0 +1,13 @@
+// double_variable_declaration_050.sol
```
In the long run, we probably need another level of directories, but it might be fine like this for now.

chriseth commented on test/libsolidity/SyntaxTestParser.cpp:
```diff
@@ -0,0 +1,77 @@
+*/
+
+#include "SyntaxTestParser.h"
+#include <fstream>
```
Please use absolute paths for all includes and sort them such that the most general includes (std) should be last and the header file belonging to the cpp file is first.

chriseth commented on test/libsolidity/SyntaxTestParser.cpp:
```diff
@@ -0,0 +1,77 @@
+namespace test
+{
+
+SyntaxTest SyntaxTestParser::parse(std::string const& _filename)
```
We usually use `using namespace std` in cpp files to make it possible to faster grasp what is happening, but it is not required.

chriseth commented on test/libsolidity/SyntaxTestParser.cpp:
```diff
@@ -0,0 +1,77 @@
+	std::string const delimiter("// ----");
+	while (std::getline(file, line))
+	{
+		if (!line.compare(0, delimiter.size(), delimiter))
```
What about `boost::startswith`? `compare` is easy to mis-use - you have to remember the meaning of the first and second argument, for example and also know what the return value means.

chriseth commented on test/libsolidity/SyntaxTestParser.cpp:
```diff
@@ -0,0 +1,77 @@
+		else
+			result.source += line + "\n";
+	}
+	while (std::getline(file, line))
```
Perhaps extract this into its own function?

chriseth commented on test/libsolidity/SyntaxTestParser.cpp:
```diff
@@ -0,0 +1,77 @@
+		auto it = line.begin();
+
+		while (it != line.end() && (std::isspace(*it) || *it == '/'))
+			it++;
```
`++it`?

chriseth commented on test/libsolidity/SyntaxTestParser.cpp:
```diff
@@ -0,0 +1,77 @@
+
+		if (it == line.end()) continue;
+
+		auto type_begin = it;
```
Please use camelCase instead of snake_case.

chriseth commented on test/libsolidity/SyntaxTestParser.cpp:
```diff
@@ -0,0 +1,77 @@
+		// skip colon
+		if (it != line.end()) it++;
+
+		while (it != line.end() && std::isspace(*it))
```
This could be its own function.

chriseth commented on test/libsolidity/SyntaxTestHelper.cpp:
```diff
@@ -0,0 +1,159 @@
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <boost/test/unit_test.hpp>
```
The boost includes should be last.

chriseth commented on test/libsolidity/SyntaxTestHelper.cpp:
```diff
@@ -0,0 +1,159 @@
+namespace test
+{
+
+class SyntaxTester : public AnalysisFramework
```
No space in front of `:`

chriseth commented on test/libsolidity/SyntaxTestHelper.h:
```diff
@@ -0,0 +1,31 @@
+namespace test
+{
+
+void registerSyntaxTests();
```
This looks a little lonely. What about moving the SyntaxTester class here and making this a static function?

chriseth commented on test/libsolidity/SyntaxTestHelper.cpp:
```diff
@@ -0,0 +1,159 @@
+	{
+		auto search_path =
+		{
+			// assuming call from source tree root
```
I have the feeling that with such logic, it is a bit too easy to accidentally find the wrong directory. Perhaps such detection logic should rather be inside a shell script that runs the test.

chriseth commented on test/libsolidity/SyntaxTestHelper.cpp:
```diff
@@ -0,0 +1,159 @@
+	if (boost::filesystem::is_directory(fullpath))
+	{
+		boost::unit_test::test_suite* sub_suite = BOOST_TEST_SUITE(_path.filename().string());
+		for (auto &entry : boost::filesystem::directory_iterator(fullpath))
```
no space in front of `&` and `:`, but space after `&`

chriseth replied:
Oh and why not `const& entry`?

chriseth commented on test/libsolidity/SyntaxTestHelper.cpp:
```diff
@@ -0,0 +1,159 @@
+	else
+	{
+		_suite.add(boost::unit_test::make_test_case(
+			[fullpath] { syntaxTester.runSyntaxTest(syntaxTestParser.parse(fullpath.string())); },
```
Does this caputure both `syntaxTester` and `syntaxTestParser`? This looks a little weird to me, to be honest.

chriseth replied:
Why do they need to be static anyway? In other boost tests, the full fixture is re-created for each test case, I think we should follow that pattern.

ekpyron replied:
Although I'll change this, as already discussed earlier, I read up on lambda expressions and capturing in such cases; from http://en.cppreference.com/w/cpp/language/lambda:
> A lambda expression can use a variable without capturing it if the variable
> - is a non-local variable or has **static** or thread local storage duration (in which case the variable cannot be captured), or 
> - ...

So using the function scope static variables inside the lambda without capturing them is in fact conformant :-).

Which makes sense, since their lifetime will only end after main() anyways.

ekpyron replied:
Actually I just also read up on boost test fixtures and realized that AnalysisFramework in SolidityNameAndTypeResolution.cpp is defined as a test suite fixture, which is in fact **not** recreated for every test, but there is only one instance for the whole test suite.

From http://www.boost.org/doc/libs/1_59_0/libs/test/doc/html/boost_test/tests_organization/fixtures/case.html#boost_test.tests_organization.fixtures.case.fixture_for_a_complete_subtree:
> In other words unless reset by another BOOST_FIXTURE_TEST_SUITE or BOOST_FIXTURE_TEST_CASE definition the same fixture is assigned to all test cases of a test suite, including ones that belong to the sub test suites. 

So following the existing pattern would in fact be to use static variables here, so given that I tend to leave it at that... any thoughts?

ekpyron replied:
[removed last comment since I was mistaken - the boost tests in fact call the constructor and destructor of the fixture for every test case, so I will change this to be consistent with the existing pattern]

chriseth commented on test/libsolidity/SyntaxTestHelper.cpp:
```diff
@@ -0,0 +1,159 @@
+		auto errorList = parseAnalyseAndReturnError(_test.source, true, true, true).second;
+
+		// checks for presence of errors and warnings with multiplicities
+		// could be modified to enforce the precise order as well
```
Yes, I think we should enforce the order.

chriseth commented on test/libsolidity/SyntaxTestHelper.cpp:
```diff
@@ -0,0 +1,159 @@
+			std::string msg = errorMessage(*error);
+			bool found = false;
+			for (auto it = expectations.begin(); it != expectations.end(); ++it)
+				if (msg.find(it->second) != std::string::npos && error->typeName() == it->first)
```
I think we should change this to check for equality, not for substring.

chriseth commented on test/libsolidity/SyntaxTestHelper.cpp:
```diff
@@ -0,0 +1,159 @@
+
+		if (!unexpectedErrors.empty())
+		{
+			std::string msg = "Unexpected error(s):\n";
```
Can you combine this and the one below into a single error message?

ekpyron commented on scripts/isolate_tests.py:
```diff
@@ -91,5 +91,8 @@ def write_cases(tests):
+                write_cases([open(path, "r").read()])
+              else:
+                cases = extract_test_cases(path)
             write_cases(cases)
```
There are quite a number of contracts in test/compilationTests/ already - so far it looks to me that they had not been processed in isolate_tests.py, but they are now.
I haven't found the place where test/compilationTests/ is processed so far - it may be that this needs adjustment to avoid double-testing those scripts.

chriseth replied:
Those are run via `cmdlineTests.sh`, and they are actually multi-file codebases, so it does not make sense to extract them into individual files, because the import each other

chriseth commented on test/libsolidity/SyntaxTestHelper.cpp:
```diff
@@ -0,0 +1,153 @@
+	else
+	{
+		_suite.add(make_test_case(
+			[fullpath] { SyntaxTester().runTest(SyntaxTestParser().parse(fullpath.string())); },
```
Somehow, your comment about fixtures and recreating the test got lost. I'm not 100% sure what to do best, but perhaps it would be a good idea to actually create a test fixture here, can we do that?

chriseth replied:
Ah sorry, now I read your comment on the deleted comment ;-)

ekpyron replied:
That comment was wrong anyways - the boost documentation is somewhat misleading there. It turns out that fixtures are in fact "shared" among all tests and there is "only one instance", but still for each individual test case the constructor and destructor of the fixture is called...
So the only difference between the current code and a boost fixture is that the boost fixture would do less memory allocations - I can look into using an actual boost fixture here, though.

ekpyron replied:
Yes - it was too late when I realized, that deleting a comment together with mail notifications may end up being confusing :-).

chriseth replied:
It's fine like this, the object is allocated on the stack anyway.

Editing comments is also fun, you follow a link and the comment says something completely different :)

chriseth commented:
The test report xml seems to contain binary data: https://circleci.com/gh/ethereum/solidity/4376

chriseth commented on scripts/isolate_tests.py:
```diff
@@ -91,5 +91,8 @@ def write_cases(tests):
             else:
-              cases = extract_test_cases(path)
+              if f.endswith(".sol"):
+                cases = [open(path, "r").read()]
```
This is weird, I would have expected the tests to fail for .sol files that import other .sol files. Can you please check?

ekpyron replied:
I'm just looking through the test scripts and tried to look up where isolate_tests.py is actually used.

The first case is cmdlineTests.sh - here the files collected from test/ are only tested using solfuzzer - however, it seems that for me solfuzzer succeeds even for entirely random input, which I suppose shouldn't be the case... still trying to figure out what's happening there.

The second case is storebytecode.sh, which ultimately calls prepare_report.py, which creates a report.txt, in which all tests with imports are reported to fail as excepted. Still trying to figure out why that doesn't result in reported test failures.

So I'm still investigating why this does not cause more problems (which seems rather weird), but ultimately it will probably make sense to exclude test/compilationTests/ in isolate_tests.py.

ekpyron replied:
storebytecode.sh seems to only be run on the travis and at the moment those tests fail earlier due to the older boost version.

chriseth replied:
solfuzzer is meant to detect crashes and internal compiler errors inside the compiler. It is meant to be run with a fuzzing library and should fail only in these cases, but the tests also just run it on specific inputs.

storebytecode is used to detect differences in compilation across different platforms. The bytecode is stored in a specific github repository and the files are compared once we have all architectures.

So all of this is fine, the question is whether we actually need it.

chriseth commented on test/libsolidity/SyntaxTestHelper.h:
```diff
@@ -0,0 +1,48 @@
+namespace test
+{
+
+class SyntaxTester: public AnalysisFramework
```
Now this file should be called `SyntaxTester.h`

chriseth commented on test/libsolidity/SyntaxTestHelper.cpp:
```diff
@@ -0,0 +1,153 @@
+using namespace boost::unit_test;
+namespace fs = boost::filesystem;
+
+namespace dev
```
We usually don't put definitions inside namespaces, only declarations. We use `using namespace dev;` and so on.

chriseth commented on test/libsolidity/SyntaxTestHelper.cpp:
```diff
@@ -0,0 +1,153 @@
+	{
+		for (size_t i = 0; i < errorList.size(); i++)
+		{
+			auto const& error = errorList[i];
```
I think this is easier to read if you do not introduce a new variable at this point. Replacing it by `errorList[i]` performs another index access, but I think that is worth it.

chriseth commented on test/libsolidity/SyntaxTestHelper.cpp:
```diff
@@ -0,0 +1,153 @@
+
+	bool errorsMatch = true;
+
+	if (errorList.size() != expectations.size ())
```
No space in front of `(`

chriseth commented on test/libsolidity/SyntaxTestHelper.cpp:
```diff
@@ -0,0 +1,153 @@
+		{
+			auto const& error = errorList[i];
+			auto const& expectation = expectations[i];
+			bool const typeMatches = error->typeName() == expectation.first;
```
It is probably better creating a struct for the type-message-pair inside SyntaxTest instead of creating these bools here.

chriseth commented on test/libsolidity/SyntaxTestHelper.cpp:
```diff
@@ -0,0 +1,153 @@
+	{
+		string msg = "Test expectation mismatch.\nExpected result:\n";
+		if (expectations.empty())
+			msg += "\tSuccess\n";
```
To be pedantic, this should be "Success without warnings", but we can also keep it like that.

ekpyron replied:
Since the result of a "Success with warning(s)" will not be nor include "Success", but just be a list of warnings, I would say "Success" should be enough for the case without warnings.

chriseth commented on test/libsolidity/SyntaxTestHelper.cpp:
```diff
@@ -0,0 +1,153 @@
+			msg += "\tSuccess\n";
+		else
+			for (auto const &expectation: expectations)
+			{
```
No braces.

chriseth commented on test/libsolidity/SyntaxTestHelper.cpp:
```diff
@@ -0,0 +1,153 @@
+		if (expectations.empty())
+			msg += "\tSuccess\n";
+		else
+			for (auto const &expectation: expectations)
```
no space in front of `&`

chriseth commented on test/libsolidity/SyntaxTestHelper.cpp:
```diff
@@ -0,0 +1,153 @@
+		if (errorList.empty())
+			msg += "\tSuccess\n";
+		else
+			for (auto &error: errorList)
```
Spacing

chriseth commented on test/libsolidity/SyntaxTestHelper.cpp:
```diff
@@ -0,0 +1,153 @@
+
+std::string SyntaxTester::errorMessage(Error const &_e)
+{
+	if (_e.comment())
```
What about `return boost::replace_all_copy(_e.comment() ? *_e.comment() : "NONE", "\n", "\\n");`?

chriseth replied:
ok, yeah, that might be too complicated :)

chriseth commented on test/libsolidity/SyntaxTestHelper.cpp:
```diff
@@ -0,0 +1,153 @@
+		return msg;
+	}
+	else
+	{
```
No braces.

chriseth commented on test/libsolidity/SyntaxTestHelper.cpp:
```diff
@@ -0,0 +1,153 @@
+	if (_e.comment())
+	{
+		std::string msg = *_e.comment();
+		boost::replace_all(msg, "\n", "\\n");
```
But perhaps at least `return boost::replace_all_copy(*_e.comment(), "\n", "\\n");` here

chriseth commented on test/libsolidity/SyntaxTestParser.cpp:
```diff
@@ -0,0 +1,107 @@
+{
+
+template<typename IteratorType>
+void skipWhitespace(IteratorType &it, IteratorType end)
```
`&` to the left

ekpyron replied:
Yeah - there's a few more of those - I'm on changing it everywhere - it's hard to drop this habit :-).

chriseth commented on test/libsolidity/SyntaxTestParser.h:
```diff
@@ -0,0 +1,58 @@
+
+struct SyntaxTestExpectation
+{
+	SyntaxTestExpectation(std::string _type, std::string _message) : type(_type), message(_message) {}
```
You can use `SyntaxTestExpectation{"type", "message"}` even without explicit constructor. If this one stays: Why didn't you use `type(std::move(_type))`?

ekpyron commented:
The travis tests fail, since an older version of boost is used there, which exposes a different API for manual test registration. Not sure whether there's a really nice way to work around that, but I'm still looking.

chriseth commented:
Ah shoot! We still need to support trusty until mid 2019, so the only option would be to install boost manually via cmake (as cpp-ethereum already does), but that could create problems especially with the emscripten build. Is it possible to make it work in both versions? I somehow have the feeling that the old version did not support multi-level trees.

ekpyron commented:
Well, if it doesn't have to be nice, then it won't be too much of a problem to support older versions of boost as well - but I think we might not get around ugly #if's distinguishing between boost versions...
I'm still checking whether there's a way around that, but if not, then at least that should definitely work
(I'm pretty sure that multi-level trees are supported already and only the API has changed).

ekpyron commented:
What's actually the current version of boost used for those builds (resp. the minimal version we have to support)?

chriseth commented:
ubuntu trusty uses 1.54, that is probably he oldest we have to support.

ekpyron committed:
Workaround for boost < 1.59.0

ekpyron commented:
Not sure why the travis test fails:

> To github.com:ethereum/solidity-test-bytecode.git
 > ! [remote rejected] master -> master (cannot lock ref 'refs/heads/master': is at 1fe670505cf6d7ff2e619d325c1294993fdc34ee but
> expected b0520d45dc8c75b38962608cbba7d7b019a2764b)
> error: failed to push some refs to 'git@github.com:ethereum/solidity-test-bytecode.git'

chriseth commented on test/libsolidity/SyntaxTester.cpp:
```diff
@@ -0,0 +1,149 @@
+	size_t line
+)
+{
+	(void)filename;
```
We generally comment out the name of the argument instead of using this workaround (it looks rather C-ish).

ekpyron committed:
Add interactive syntax test tool.

ekpyron commented:
Rebased and squashed.

Added a first version of an interactive syntax test tool.
The code still has to be revised, but the functionality is there already.

ekpyron commented on test/tools/syntaxTestTool.cpp:
```diff
@@ -0,0 +1,354 @@
+namespace po = boost::program_options;
+namespace fs = boost::filesystem;
+
+namespace color
```
WIP: put this some better place than a global namespace

ekpyron replied:
In general color coding the output improves the usability of the tool a lot, though.

chriseth replied:
Yay, colours! :)

ekpyron commented on test/tools/syntaxTestTool.cpp:
```diff
@@ -0,0 +1,354 @@
+					break;
+				case 'e':
+					cout << endl << endl;
+					(void)system((editor + " " + m_path.string()).c_str());
```
Any coding style preferences for ignoring return values as in this case?

chriseth replied:
I'm not aware of our complier settings reporting this...

ekpyron commented on test/tools/syntaxTestTool.cpp:
```diff
@@ -0,0 +1,354 @@
+int main(int argc, char *argv[])
+{
+	DisableConsoleBuffering disableConsoleBuffering;
+	(void)disableConsoleBuffering;
```
Or any coding style preference for unused variables in a case like this (scoped setup and cleanup in constructor and destructor)?

chriseth replied:
Is it marked as unused by the compiler if the constructor has side-effects?

ekpyron commented on test/CMakeLists.txt:
```diff
@@ -1,9 +1,13 @@
+add_executable(solfuzzer tools/fuzzer.cpp)
 target_link_libraries(solfuzzer PRIVATE libsolc evmasm ${Boost_PROGRAM_OPTIONS_LIBRARIES})
+
+add_executable(syntaxTestTool tools/syntaxTestTool.cpp TestHelper.cpp libsolidity/SyntaxTestParser.cpp libsolidity/AnalysisFramework.cpp)
```
Should I add ``EXCLUDE_FROM_ALL`` here? The ``syntaxTestTool`` won't be of much use for most users - in contrast to ``solfuzzer`` it won't even be used for running any tests, so maybe it makes more sense to require to explicitly build it with ``make syntaxTestTool``?

chriseth replied:
I think we should always build everything, to see when it breaks. Most users don't actually build from source.

ekpyron committed:
Refactoring; fuse SyntaxTestParser and SyntaxTester to SyntaxTest and reuse code in syntaxTestTool.

ekpyron commented:
Now includes a syntaxTestTool that can run all syntax tests interactively and allows to either edit test contracts and test expectation or to automatically update the test expectations to the actual results of the test in case of a test failure.

ekpyron committed:
Remove (void)disableConsoleBuffering to check whether this actually results in compile time errors.

chriseth commented:
Please split this pull request into three different pull requests:

 - boost test runner
 - cmake modifications
 - expectation updater

chriseth commented on test/CMakeLists.txt:
```diff
@@ -1,9 +1,13 @@
+add_executable(solfuzzer tools/fuzzer.cpp)
 target_link_libraries(solfuzzer PRIVATE libsolc evmasm ${Boost_PROGRAM_OPTIONS_LIBRARIES})
+
+add_executable(syntaxTestTool tools/syntaxTestTool.cpp TestHelper.cpp libsolidity/SyntaxTest.cpp libsolidity/AnalysisFramework.cpp)
```
This should later be expanded to semantic tests, too. What about `isoltest' (interactive soltest)?

chriseth commented:
Please update the documentation ("how to run the tests").

chriseth commented on test/libsolidity/FormattedPrinter.h:
```diff
@@ -0,0 +1,90 @@
+		template<typename T>
+		std::ostream& operator<<(T&& t)
+		{
+			return m_stream << t;
```
`std::forward`?

chriseth commented on test/libsolidity/FormattedPrinter.h:
```diff
@@ -0,0 +1,90 @@
+	class Scope
+	{
+	public:
+		Scope(std::ostream& _stream, bool const _enabled) : m_stream(_stream), m_enabled(_enabled)
```
No space left of `:`

ekpyron commented:
Split into #3707, #3708 and #3709.

ekpyron commented:
Abandoning in favour of refactored  #3707, #3708 and #3709.
